### ISSUE TITLE
Add support for async launch policy in dash/io

### ISSUE THREAD

**Issue by fmoessbauer** [dash-project/dash#202]
Concept

```cpp
using dash::io::hdf5 as dio;
dio::OutputStream os(dash::launch::async,"file.hdf5");
// [...] Some heavy calculations
os.flush();
```

The async stream internally uses futures. To avoid multithreading in IO, the store/load operations are chained and executed sequentially.

Implementation in branch `feat-hdf5-views`

**Comment by fuchsto** [comment]
Boost Asio:
http://www.boost.org/doc/libs/1_60_0/doc/html/boost_asio/reference.html

**Comment by fuchsto** [comment]
Note that async is a *launch policy*, not an *execution* policy.

http://en.cppreference.com/w/cpp/thread/async

**Comment by fmoessbauer** [comment]
The feature is implemented and works for both input and output streams. Surprisingly it even works with builds of MPI without thread support. However then, no MPI calls are allowed while performing async io operations (otherwise strange things might happen).

I also updated both OpenMPI containers to multithreaded builds of OpenMPI. Thereto OpenMPI is build from the sources instead of using the package.

@devreal: I did a quit and dirty fix in the [initialization of DART](https://github.com/dash-project/dash/compare/feat-hdf5-views#diff-dcdf0a19e9a3ed023af50e6217525116R58): When there is a thread concept in DART, this could be enhanced.

When the implementation of the other planned features in the HDF5 driver are finished, I will make a pull request. All the related stuff is in branch [feat-hdf5-views](https://github.com/dash-project/dash/compare/feat-hdf5-views).

**Comment by fmoessbauer** [comment]
A branch containing only the new docker containers is `feat-202-mpithread`

**Comment by devreal** [comment]
I'm fine with this hack for the time being. However, unless you actually check for the value of `_provided_thread_support` before you call MPI in a multi-threaded environment it does not give you much benefit. After all, you might be using an MPI library that was not compiled with support for `MPI_THREAD_MULTIPLE`. From what I know, OpenMPI uses thread-multiple by default if configured appropriately.

**Comment by fmoessbauer** [comment]
@devreal : Give me a hint when you have finished related changes in `bug-118-threadsafety`, then I will replace my workaround with your implementation.

**Comment by fmoessbauer** [comment]
Implemented in PR #268
